modules/personalize_taxonomy_context/plugins/visitor_context/TaxonomyContext.inc

```diff
@@ -0,0 +1,93 @@
+<?php
+
+/**
+ * @file
+ * Provides a visitor context plugin for taxonomy.
+
+ */
+
+class TaxonomyContext extends PersonalizeContextBase {
+  /**
+   * Implements PersonalizeContextInterface::create().
+   */
+  public static function create(PersonalizeAgentInterface $agent = NULL, $selected_context = array()) {
+    return new self($agent, $selected_context);
+  }
+
+  /**
+   * Implements PersonalizeContextInterface::getOptions().
+   */
+  public static function getOptions() {
+    $vocabularies = taxonomy_get_vocabularies();
+    $options = array();
+    foreach ($vocabularies as $vocabulary) {
+      $machine_name = $vocabulary->machine_name;
+      $options[$machine_name] = array(
+        'name' => $vocabulary->name,
+        'group' => 'Taxonomy',
+        'id' => $vocabulary->vid,
+      );
+    }
+    return $options;
+  }
+
+  /**
+   * Implements PersonalizeContextInterface::getPossibleValues().
+   */
+  public function getPossibleValues($limit = FALSE) {
+    $possible_values = array();
+    $options = $this->getOptions();
+    $free_tagging_vocabulary_machine_names = SELF::getFreeTaggingVocabulariesMachineNames();
```

Comment:
I've spent a LONG time trying to find a definitive way to collect free-tagging vocabularies, but as the [doc](https://www.drupal.org/node/23405) says (and pretty much nowhere else), there is no db column or anything that defines a field's free-tagging-ness. So far, the criteria is still ```term-reference``` + ```auto-complete```.

Choosing a single value does not solve the problem I think ... 
1. When doing fixed targeting, we are still looking at potentially hundreds even thousands of terms for free-taggings at dropdown, and it is definitely bad UX as is. 
2. We need to send all terms of the selected vocabulary, Web needs it (Steve mentioned it on one of the morning standups). 
3. Most importantly, the reason why I think allowing free-tagging vocabularies on fixed targeting is a bad idea, because again it promotes sloppy and ineffective personalizations.

I don't think this is too bad. getPossibleValues() is called only by limited admin pages, and a few unoptimized query shouldn't harm. As long as the content pages are optimized and fast, I am not worried. If you ask me, I feel this is handled as good as I can at this point.